Adding metrics for aws-admission-controller

njuettner opened the pull request:
Adding prometheus metrics to get some insights from our aws-admission-controller

Metrics:
```
* Total number of errors
* Total number of successful requests
* Duration of requests
* Total number of invalid requests
* Total number of rejected requests
* Total number of requests
```

Example:

```
# HELP aws_admission_controller_webhook_duration_requests Duration of requests
# TYPE aws_admission_controller_webhook_duration_requests histogram
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="mutating",le="0.005"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="mutating",le="0.01"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="mutating",le="0.025"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="mutating",le="0.05"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="mutating",le="0.1"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="mutating",le="0.25"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="mutating",le="0.5"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="mutating",le="0.75"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="mutating",le="1"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="mutating",le="1.25"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="mutating",le="1.5"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="mutating",le="2"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="mutating",le="2.5"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="mutating",le="5"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="mutating",le="10"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="mutating",le="+Inf"} 1
aws_admission_controller_webhook_duration_requests_sum{resource="awscontrolplane",webhook="mutating"} 1.2297e-05
aws_admission_controller_webhook_duration_requests_count{resource="awscontrolplane",webhook="mutating"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="validating",le="0.005"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="validating",le="0.01"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="validating",le="0.025"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="validating",le="0.05"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="validating",le="0.1"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="validating",le="0.25"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="validating",le="0.5"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="validating",le="0.75"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="validating",le="1"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="validating",le="1.25"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="validating",le="1.5"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="validating",le="2"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="validating",le="2.5"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="validating",le="5"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="validating",le="10"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="awscontrolplane",webhook="validating",le="+Inf"} 1
aws_admission_controller_webhook_duration_requests_sum{resource="awscontrolplane",webhook="validating"} 5.871e-06
aws_admission_controller_webhook_duration_requests_count{resource="awscontrolplane",webhook="validating"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="g8scontrolplane",webhook="mutating",le="0.005"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="g8scontrolplane",webhook="mutating",le="0.01"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="g8scontrolplane",webhook="mutating",le="0.025"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="g8scontrolplane",webhook="mutating",le="0.05"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="g8scontrolplane",webhook="mutating",le="0.1"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="g8scontrolplane",webhook="mutating",le="0.25"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="g8scontrolplane",webhook="mutating",le="0.5"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="g8scontrolplane",webhook="mutating",le="0.75"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="g8scontrolplane",webhook="mutating",le="1"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="g8scontrolplane",webhook="mutating",le="1.25"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="g8scontrolplane",webhook="mutating",le="1.5"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="g8scontrolplane",webhook="mutating",le="2"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="g8scontrolplane",webhook="mutating",le="2.5"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="g8scontrolplane",webhook="mutating",le="5"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="g8scontrolplane",webhook="mutating",le="10"} 1
aws_admission_controller_webhook_duration_requests_bucket{resource="g8scontrolplane",webhook="mutating",le="+Inf"} 1
aws_admission_controller_webhook_duration_requests_sum{resource="g8scontrolplane",webhook="mutating"} 1.0491e-05
aws_admission_controller_webhook_duration_requests_count{resource="g8scontrolplane",webhook="mutating"} 1
# HELP aws_admission_controller_webhook_rejected_requests Total number of rejected requests
# TYPE aws_admission_controller_webhook_rejected_requests counter
aws_admission_controller_webhook_rejected_requests{resource="awscontrolplane",webhook="validating"} 1
# HELP aws_admission_controller_webhook_requests Total number of requests
# TYPE aws_admission_controller_webhook_requests counter
aws_admission_controller_webhook_requests{resource="awscontrolplane",webhook="mutating"} 1
aws_admission_controller_webhook_requests{resource="g8scontrolplane",webhook="mutating"} 1
# HELP aws_admission_controller_webhook_successful_requests Total number of successful requests
# TYPE aws_admission_controller_webhook_successful_requests counter
aws_admission_controller_webhook_successful_requests{resource="awscontrolplane",webhook="mutating"} 1
aws_admission_controller_webhook_successful_requests{resource="g8scontrolplane",webhook="mutating"} 1
```

**Note:** every admission request will go through mutating **and** validating handler for the each resource (e.g. AWSControlPlane), meaning when there's a webhook configuration for mutation and validation it will be counted against both. For that I added labels for each resource and each webhook (validating or mutating) to make it more visible.

xh3b4sd commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,56 @@
+	InternalError = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: MetricNamespace,
+		Subsystem: metricSubsystem,
+		Name:      "internal_error",
```
There are naming conventions on how to name labels and metrics. See https://prometheus.io/docs/practices/naming. Let's try to follow community conventions as much as it makes sense.

anvddriesch commented:
It looks good to me. (Not sure about naming - I agree we should follow conventions)
If the metrics don't show up yet, we may still have to add the scrape annotation.
Also, since you mention the mutating and validating webhooks. For my information, do we make sure that the validation step is always second?

njuettner commented:
Ok I changed it a bit, I hope this aligns a bit better with the convention.

@anvddriesch IIRC we added scraping in https://github.com/giantswarm/aws-admission-controller/pull/60, but let me know if something is missing. 
Validation is always ensured as the last step before persisting into etcd. So yes it the case.

anvddriesch reviewed (approved):
Thank you for explaining! :) And no, I don't think anything is missing. If the metrics show up in prometheus it's all good.
